Wrappers/Python/cil/optimisation/functions/KullbackLeibler.py

```diff
@@ -1,31 +1,239 @@
+    definition that is used in PET reconstruction in the `SIRF <https://github.com/SyneRBI/SIRF>`_ software.
+    
+       
+    Note
+    ----
+    The default implementation uses the build-in function `kl_div <https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.kl_div.html>`_ from scipy.
+    The methods of the :class:`.KullbackLeibler` are accelerated provided that `numba <https://numba.pydata.org/>`_ library is installed.
+           
+    Examples
+    --------
+    >>> from cil.optimisation.functions import KullbackLeibler 
+    >>> from cil.framework import ImageGeometry   
+    >>> ig = ImageGeometry(3,4)
+    >>> data = ig.allocate('random')
+    >>> F = KullbackLeibler(b = data)
+                        
+    """     
+
+    def __new__(cls, b, eta = None, mask = None, backend = None):
+
+        # default backend numba
+        cls.backend = backend
+        
+        if cls.backend == 'numba':
+
+            if not has_numba:
+                raise ValueError(" Numba is not installed. ")
+            else:          
+                logging.info(" Numba backend is used. ")      
+                return super(KullbackLeibler, cls).__new__(KullbackLeibler_numba)
+        else:
+            logging.info(" Numpy backend is used. ") 
+            return super(KullbackLeibler, cls).__new__(KullbackLeibler_numpy)
+
+    def __init__(self, b, eta = None, mask = None, backend = None):
+
+        self.b = b
+        self.eta = eta
+        self.mask = mask
+
+        if self.eta is None:
+            self.eta = self.b * 0.0
+
+        if (self.b.as_array() < 0).any():            
+            raise ValueError(" Input data should be non-negative.")         
+
+        if KullbackLeibler.backend == 'numba':
+
+            self.b_np = self.b.as_array()
+            self.eta_np = self.eta.as_array()
+
+            if mask is not None:
+                self.mask = mask.as_array()
+                        
+        super(KullbackLeibler, self).__init__(L = None) 
+
+class KullbackLeibler_numpy(KullbackLeibler):
+       
+    def __call__(self, x):
+
+        r"""Returns the value of the KullbackLeibler function at :math:`(b, x + \eta)`.
+
+        Note
+        ----
+        To avoid infinity values, we consider only pixels/voxels for :math:`x+\eta\geq0`.
+        """
+
+        tmp_sum = (x + self.eta).as_array()
+        ind = tmp_sum >= 0
+        tmp = scipy.special.kl_div(self.b.as_array()[ind], tmp_sum[ind])             
+        return numpy.sum(tmp) 
+
+    def gradient(self, x, out = None):
+
+        r"""Returns the value of the gradient of the KullbackLeibler function at :math:`(b, x + \eta)`.                
+        
+        :math:`F'(b, x + \eta) = 1 - \frac{b}{x+\eta}`
+        
+        Note
+        ----
+        To avoid inf values, we :math:`x+\eta>0` is required. 
+        
+        """         
+
+        tmp_sum_array = (x + self.eta).as_array()
+
+        if out is None:   
```

Comment:
should_return=False
if out is None:
    out = x.add(self.eta)
    should_return=True
else:
    x.add(self.eta,out=out)

out_np= out.as_array()
out_np[out_np>0] = 1 - self.b.as_array()[out_np>0]/out_np[out_np>0]

out.fill(out_np)

if should_return:
    return out

Reply:
Added in the `gradient` method of `KullbackLeibler` with numpy backend.